File: packages/genai/lib/models/a2ui.dart
import 'dart:convert';

/// Parses Agent-to-UI (A2UI) JSONL payloads into components and data model.
///
/// The A2UI protocol streams newline-delimited JSON events:
///   createSurface       — surface metadata (id, title, layout)
///   updateComponents    — component tree nodes
///   updateDataModel     — data bindings
///   closeSurface        — surface teardown (ignored, no state needed)
class A2UIParser {
  static bool isA2UIPayload(String body) {
    for (final line in body.split('\n')) {
      final t = line.trim();
      if (t.isEmpty) continue;
      try {
        final json = jsonDecode(t);
        if (json is Map &&
            (json.containsKey('createSurface') ||
                json.containsKey('updateComponents'))) {
          return true;
        }
      } catch (_) {}
    }
    return false;
  }

  static ({
    Map<String, dynamic> components,
    Map<String, dynamic> dataModel,
    String? surfaceTitle,
  })? parse(String body) {
    final Map<String, dynamic> components = {};
    final Map<String, dynamic> dataModel = {};
    String? surfaceTitle;

    for (final line in body.split('\n')) {
      final t = line.trim();
      if (t.isEmpty) continue;
      try {
        final json = jsonDecode(t);
        if (json is! Map<String, dynamic>) continue;

        if (json.containsKey('createSurface')) {
          final msg = json['createSurface'] as Map<String, dynamic>?;
          surfaceTitle = msg?['title'] as String?;
        }

        if (json.containsKey('updateComponents')) {
          final msg = json['updateComponents'] as Map<String, dynamic>?;
          for (final c in (msg?['components'] as List? ?? [])) {
            final comp = c as Map<String, dynamic>;
            final id = comp['id'] as String?;
            if (id != null) components[id] = comp;
          }
        }

        if (json.containsKey('updateDataModel')) {
          final msg = json['updateDataModel'] as Map<String, dynamic>?;
          final path = msg?['path'] as String?;
          if (path != null && path.startsWith('/')) {
            dataModel[path.substring(1)] = msg?['value'];
          }
        }
      } catch (_) {}
    }

    return components.isEmpty
        ? null
        : (
            components: components,
            dataModel: dataModel,
            surfaceTitle: surfaceTitle,
          );
  }
}

File: api/a2ui.py
"""
GET /a2ui

Returns an agent-to-UI JSONL payload — newline-delimited JSON commands that
the Flutter client processes sequentially to build a live UI surface.

Command sequence:
  1. createSurface    — initialise the canvas (id, title, layout)
  2. updateComponents — declare the full component tree in one batch
  3. updateDataModel  — push live metric values into data-bound nodes

The response MIME type is text/plain because each line is a self-contained
JSON object; clients split on newlines and parse each line individually.

Used by the APIDash PoC to demonstrate Generative UI / agent-driven rendering.
"""
from http.server import BaseHTTPRequestHandler
import sys
import os

# Make the parent api/ directory importable so _data.py can be found
sys.path.insert(0, os.path.dirname(__file__))
from _data import A2UI_JSONL


class handler(BaseHTTPRequestHandler):
    def do_OPTIONS(self):
        # Handle CORS preflight requests from browser clients
        self.send_response(200)
        self._cors()
        self.end_headers()

    def do_GET(self):
        # Return the pre-built JSONL string as plain text
        # Clients split on '\n' and parse each line as a separate JSON command
        body = A2UI_JSONL.encode()
        self.send_response(200)
        self.send_header('Content-Type', 'text/plain')
        self._cors()
        self.end_headers()
        self.wfile.write(body)

    def _cors(self):
        # Allow cross-origin requests from the Flutter web client and APIDash UI
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'GET, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type')
